src/daft-parquet/src/file.rs

```diff
@@ -403,159 +411,162 @@ impl ParquetFileReader {
+                                        Arc::new(|_, _| true),
+                                        Self::MAX_PAGE_SIZE,
+                                    )
+                                    .with_context(|_| {
+                                        UnableToCreateParquetPageStreamSnafu::<String> {
+                                            path: uri.clone(),
+                                        }
+                                    })?;
+                                let page_stream = streaming_decompression(compressed_page_stream);
+                                let pinned_stream = Box::pin(page_stream);
+                                decompressed_iters
+                                    .push(StreamIterator::new(pinned_stream, rt_handle.clone()));
                             }
-                        });
-                        Ok(())
-                    })
+                            let arr_iter = column_iter_to_arrays(
+                                decompressed_iters,
+                                ptypes.iter().collect(),
+                                field,
+                                Some(chunk_size),
+                                num_rows,
+                                num_values,
+                            )?;
+                            Ok(arr_iter)
+                        })
+                    });
+
+                    let arr_iters = try_join_all(arr_iter_handles)
+                        .await
+                        .context(JoinSnafu { path: uri.clone() })?
+                        .into_iter()
+                        .collect::<DaftResult<Vec<_>>>()?;
+
+                    DaftResult::Ok(arr_iters)
                 });
+                (row_range, chunk_iter_task)
+            });
+
+        let compute_runtime = get_compute_runtime();
+        let uri = self.uri.clone();
+        let parquet_task = compute_runtime.spawn(async move {
+            let mut table_tasks = Vec::with_capacity(chunk_iter_handles.len());
+            for ((row_range, chunk_iter_handle), output_sender) in chunk_iter_handles.zip(senders) {
+                // We want to ensure that the channel capacity can hold one morsel worth of data for better deserialization performance.
+                let channel_size = {
+                    let chunks_per_morsel = max(PARQUET_MORSEL_SIZE / chunk_size, 1);
+                    let max_num_chunks = max(row_range.num_rows / chunk_size, 1);
+                    min(chunks_per_morsel, max_num_chunks)
+                };
 
-        let _ = try_join_all(table_iter_handles)
-            .await
-            .context(JoinSnafu { path: self.uri })?
-            .into_iter()
-            .collect::<DaftResult<Vec<_>>>()?;
+                let chunk_iter = chunk_iter_handle
+                    .await
+                    .context(JoinSnafu { path: uri.clone() })??;
+
+                let permit = semaphore.clone().acquire_owned().await.unwrap();
```

Comment:
discussed offline, this makes sense for now but for the bulk reader we may want to share this permit across files